[0.目录](../0.目录.md)
# 4.4 处理缺陷
大多数错误处理运算符仅处理错误而不是缺陷。 例如，如果我们回到上一章的foldM的签名，我们会发现没有理由处理Throwable类型的缺陷：

```
  final case class ZIO[-R, +E, +A](run: R => Either[E, A]) {
    self =>
    def foldM[R1 <: R, E1, B](
      failure: E => ZIO[R1, E1, B],
      success: A => ZIO[R1, E1, B]
    ): ZIO[R1, E1, B] =
      ZIO(r => self.run(r).fold(failure, success).run(r))
  }
```

这种哲学是一个很好的默认设置，因为在大多数情况下，处理缺陷是没有意义的。
缺陷表示无法预期或无法恢复的故障。
因此，在大多数情况下，我们不想通过指定如何处理我们从未期望过并且可能无法恢复的未知故障来使我们的错误处理操作的签名进一步复杂化。
但是，在某些情况下，我们可能还要处理故障以外的缺陷。
例如，如果我们正在为应用程序实现日志记录，则尽管无法采取任何措施从缺陷中恢复，我们可能仍希望记录缺陷以保留信息。

在我们的应用中，处理缺陷在不同层之间的边缘处尤其重要。
让我们考虑报告生成应用程序的上述示例，该示例使用控制台应用程序允许用户自定义报告设置。
在控制台应用程序级别，无法处理IOException，因此这将被视为缺陷，并且控制台应用程序将因IOException而死亡。
另一方面，报告生成应用程序可以将此视为失败，并通过使用默认设置生成报告来处理它。
ZIO提供了一个独立的运营商系列，可为您提供这种灵活性。
从概念上讲，这些运算符中最基本的是沙箱，它具有以下特征：
```
trait ZIO[-R, +E, +A] {
    def sandbox: ZIO[R, Cause[E], A]
}
```

对effect进行沙盒处理可以在我们的effect的错误通道中揭示失败的全部原因。这使我们可以使用所有常规的错误处理运算符来处理Cause。
完成后，如果我们想将完整的Cause淹没到effect中，那么我们只需要处理类型错误，就可以使用unsandbox操作，它是沙箱的逆过程。

```
trait ZIO[-R, +E, +A] {
    def unsandbox[E1](implicit ev: E <:< Cause[E1]): ZIO[R, E1, A]
}
```

sandbox和unsandbox一样强大，足以实施我们可能想要处理全部故障原因的任何操作。但是，ZIO还提供了各种便利方法，这些方法旨在简化特定用例。
foldCauseM是最通用的运算符之一。
这就像我们之前讨论过的foldM一样，但是现在错误情况包含了失败的全部原因：

```
  final case class ZIO[-R, +E, +A](run: R => Either[Cause[E], A]) {
    self =>
    def foldCauseM[R1 <: R, E1, B](
      failure: Cause[E] => ZIO[R1, E1, B],
      success: A => ZIO[R1, E1, B]
    ): ZIO[R1, E1, B] =
      ZIO(r => self.run(r).fold(failure, success).run(r))
  }
```

就像foldM可用于实现各种错误处理运算符一样，foldCauseM可以用于实现许多更具体的运算符来处理故障的全部原因。



